realm/realm-library/src/androidTest/java/io/realm/OrderedRealmCollectionIteratorTests.java

```diff
@@ -0,0 +1,964 @@
+        final CountDownLatch bgDone = new CountDownLatch(1);
+        Iterator<AllJavaTypes> it = collection.iterator();
+        it.next();
+
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Realm realm = Realm.getInstance(OrderedRealmCollectionIteratorTests.this.realm.getConfiguration());
+                appendElementToCollection(realm, collectionClass);
+                realm.close();
+                bgDone.countDown();
+            }
+        }).start();
+        TestHelper.awaitOrFail(bgDone);
+
+        realm.refresh();
+        switch (collectionClass) {
+            case UNMANAGED_REALMLIST:
+                assertEquals(TEST_SIZE, collection.size());
+                break;
+
+            case MANAGED_REALMLIST:
+            case REALMRESULTS:
+                assertEquals(TEST_SIZE + 1, collection.size());
+                break;
+
+            default:
+                fail("Unknown class: " + collectionClass);
+        }
+    }
+
+    // TODO Remove once waitForChange is introduced
+    @Test
+    @UiThreadTest
+    public void listIterator_refreshWhileIterating_looper() {
+        final CountDownLatch bgDone = new CountDownLatch(1);
+        Iterator<AllJavaTypes> it = collection.iterator();
+        it.next();
+
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Realm realm = Realm.getInstance(OrderedRealmCollectionIteratorTests.this.realm.getConfiguration());
+                appendElementToCollection(realm, collectionClass);
+                realm.close();
+                bgDone.countDown();
+            }
+        }).start();
+        TestHelper.awaitOrFail(bgDone);
+
+        realm.refresh();
+        switch (collectionClass) {
+            case MANAGED_REALMLIST:
+            case UNMANAGED_REALMLIST:
+            case REALMRESULTS:
+                assertEquals(TEST_SIZE, collection.size());
+                break;
+
+            default:
+                fail("Unknown class: " + collectionClass);
+        }
+    }
+
+    @Test
+    public void listIterator_deleteManagedObjectIndirectly() {
+        realm.beginTransaction();
+        Iterator<AllJavaTypes> it = collection.iterator();
+        it.next();
+        it.next().deleteFromRealm();
+        realm.commitTransaction();
+
+
+        switch (collectionClass) {
+            case MANAGED_REALMLIST:
+                assertEquals(TEST_SIZE - 1, collection.size());
+                break;
+            case UNMANAGED_REALMLIST:
+            case REALMRESULTS:
+                assertEquals(TEST_SIZE, collection.size());
+                break;
+        }
+        it = collection.listIterator();
+        it.next();
+        AllJavaTypes types = it.next(); // Iterator can still access the deleted object
+
+        assertFalse(collectionClass == CollectionClass.MANAGED_REALMLIST ? !types.isValid() : types.isValid());
```

Comment:
`assertTrue(collectionClass == CollectionClass.MANAGED_REALMLIST ? types.isValid() : !types.isValid());`
for easier reading